### ISSUE TITLE
Organism.dplyr

### ISSUE THREAD

**Issue by yubocheng** [Bioconductor/Contributions#263]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/Bioconductor/Organism.dplyr

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Subversion][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/source-control/
[6]: http://bioconductor.org/developers/how-to/git-mirrors/

**Comment by lawremi** [comment]
I'm concerned about how this promotes low-level data access, directly on the schema. Isn't that something that should be abstracted? I like the high-level filtering. That stuff should probably be pushed up to GenomicFeatures or beyond. Any potential for integration with ensembldb?

**Comment by yubocheng** [comment]
This package doesn’t do much about low-level data access. It combined ‘OrgDb’ and ‘TxDb’ into one database, so user could use dplyr to query through information at one place. In schema folder, there is one SQL file for each organism/OrgDb to create new database, if there is any change of OrgDb data structure, the corresponding SQL file needs to be changed as well. There is no plan to for integration with ensembldb at this point.

**Comment by lawremi** [comment]
I'm just saying that the user code is depending on the schema, without any abstraction. Integration with ensembldb would be a good idea.

**Comment by lawremi** [comment]
While it's nice to provide low-level access, one could also imagine dplyr workflows operating on a single table view of the database. The view would have one row per transcript, with all of its information (coords, id, gene id, cds start/end, etc).

**Comment by mtmorgan** [comment]
@lawremi At first we used a  SQL VIEW onto the original TxDb and org schemas, and the abstraction was at the level of the VIEW; this is as flexible as a traditional software abstraction, with tweaks to the VIEW accommodating changes in interface or implementation. The pure VIEW approach turned out not to be performant, especially for the between-database joins that are required to combine TxDb and org components. For this reason, we switched to re-creating tables at run-time, with the schema exposed via dplyr. A VIEW could be inserted again as an abstraction layer on the new table structures, with dplyr seeing the VIEW, if an abstraction were required.

Maybe what you're saying is that you'd really like us to structure the data differently (tx centric?) or to use VIEWs on the new tables to provide convenient access to something like `left_join(tbl(src, "tx"), tbl(src, "id"))` ?

**Comment by lawremi** [comment]
It seems like people are able to understand single, denormalized tables easier than multi-table schema. As a first shot, you could use the GFF3 model.  I think one could convert that into many useful forms using basic dplyr verbs. Basically, GFF3 is a tidy TxDb.

**Comment by jotsetung** [comment]
Regarding integration of `ensembldb`, let me know if I can give a hand there.

**Comment by jotsetung** [comment]
The `Filter` classes look also pretty much like the `ensembldb` filters - wouldn't it be better to import them from `ensembldb` instead of implementing them - also regarding potential clashes when `ensembldb` is loaded?

**Comment by jotsetung** [comment]
@lawremi I agree that less tables would be easier for users (especially to start with). And what about having a table for *genes*, *transcripts* and *exons* (eventually proteins etc) each of them also providing their genomic coordinates? This would already reduce the number of tables (no *ranges* tables). I wouldn't put all data into a single table since that would have an impact on performance (especially joins will be terribly slow) - also, in many cases users might not be really interested in the exons when they search for genes.

**Comment by mtmorgan** [comment]
I started issues on the Organism.dplyr package for 'appropriate tables exposed to the user' and 'filters', so the discussion is more hierarchical. We should carry on the discussion there.

**Comment by vobencha** [comment]
1) BasicFilter man page
- The title is not very descriptive:
```
\title{Functions for filter objects}
```
I'd clarify in \title or \description that these filters are intended for extractor functions that operate on src_organism objects.

- Typo:
```
\description{
These functions are used to create filters for genomic extrators.
}
```
- I'd add more to \details to clarify the 'columns' are columns of the merged database. Maybe show how they can access the columns ... or refer them to ?`columns,src_organism-method`.
```
\details{
All these filters except \code{GRangesFilter()} extend \code{BasicFilter}
class. Each filter takes value(s) from the corresponding column. For
example, \code{AccnumFilter()} takes value of accession number(s), which
come from column \code{accnum}.
```
2) select man page
Assuming the examples are wrapped in \dontrun due to time constraints? I see select() is evaluated in the vignette so that's probably good enough?

3) scr_organism man page
Should we use terms OrgDb and TxDb instead of org and txdb?
```
\title{Create a sqlite database from an org package and a txdb package}
```
4) utils man page
Typo:
```
\description{
These functions are primarily for illustrating
funtionality.
```
5) Should we add an \author field on the man pages with Yubo's name?

I like how the vignette has been re-organized into the '3 interfaces'. This makes the goal of the package much more clear.

Val

**Comment by vobencha** [comment]
Looks good.
